aten/src/ATen/mps/MPSDevice.mm

```diff
@@ -3,26 +3,72 @@
 #include <c10/util/CallOnce.h>
 
 #include <ATen/mps/MPSDevice.h>
+#include <ATen/mps/IndexKernels.h>
 
 namespace at {
 namespace mps {
 
 static std::unique_ptr<MPSDevice> mps_device;
 static c10::once_flag mpsdev_init;
 
+static inline MTLLanguageVersion getMetalLanguageVersion(const id<MTLDevice>& device) {
+  // MPS Advanced Indexing needs at least Metal 2.0 (support for Argument Buffers and function constants)
+  MTLLanguageVersion languageVersion;
+
+#if defined(__MAC_13_0) && __MAC_OS_X_VERSION_MIN_REQUIRED >= __MAC_13_0
+  languageVersion = MTLLanguageVersion3_0;
+#elif defined(__MAC_12_0) && __MAC_OS_X_VERSION_MIN_REQUIRED >= __MAC_12_0
+  languageVersion = MTLLanguageVersion2_4;
+#elif
+  #error "Metal is not available on the current platform."
+#endif
```

Comment:
Code as its currently written will fail to compile on MacOS 14.0, when it will be eventually released.
And hardcoding version would prevent as well prevent us from encountering a runtime errors, if newer MTL Language Standards would make some incompatible changes. 
(And applied compiler optimizations are independent of MTLLanguageVersion, are they?)
```suggestion
  // MPS Advanced Indexing needs at least Metal 2.0 (support for Argument Buffers and function constants)
  // host_name attribute needs at least Metal 2.2
  MTLLanguageVersion languageVersion = MTLLanguageVersion2_2;
```

Reply:
> Code as its currently written will fail to compile on MacOS 14.0, when it will be eventually released.

This shouldn't fail to compile on newer macOS. MacOS Ventura is 13.0 and according to the macro logic, it would fall into this macro logic:
```
#if defined(__MAC_13_0) && __MAC_OS_X_VERSION_MIN_REQUIRED >= __MAC_13_0
  languageVersion = MTLLanguageVersion3_0;
```
Even for higher numbers it would fall into the same macro logic (this `__MAC_OS_X_VERSION_MIN_REQUIRED >= __MAC_13_0` is checking for the number to be >= 13, not strictly equal to 13)

Reply:
Hardcoding to 2_2 as per suggestion.